From d13a25cc8270486bddabcd07114906b625162212 Mon Sep 17 00:00:00 2001
From: Yoni Farhadia <yonif@marvell.com>
Date: Sun, 27 Oct 2013 10:54:04 +0200
Subject: [PATCH 1069/1825] pp2: alp: MAC to SGMII support

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 5019f06818d4c1bc505343d32c41ee8403bb66f6

        - Enable MAC inband AN in case of SGMII
        - Enable MAC PCS in case of SGMII
        - Set plat_data flag for SGMII in case mvBoardIsPortSgmii is true

Change-Id: I9f8907033cc0f6f603e43efb1d721c1ae6a75b8c
Signed-off-by: Yoni Farhadia <yonif@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/3859
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Dmitri Epshtein <dima@marvell.com>
Reviewed-by: Igor Patrik <igorp@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 arch/arm/mach-avantalp/core.c                      |    5 +++++
 .../arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacApi.c |    5 +++--
 .../plat-armada/mv_hal/pp2/gmac/mvEthGmacRegs.h    |    3 +++
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-avantalp/core.c b/arch/arm/mach-avantalp/core.c
index a4403ad..c1e15dc 100644
--- a/arch/arm/mach-avantalp/core.c
+++ b/arch/arm/mach-avantalp/core.c
@@ -612,6 +612,11 @@ static void mv_pp2_giga_pdev_register(struct platform_device *pdev)
 	if (mvBoardIsPortLoopback(port))
 		plat_data->flags |= MV_PP2_PDATA_F_LB;
 
+	if (mvBoardIsPortInSgmii(port))
+		plat_data->flags |= MV_PP2_PDATA_F_SGMII;
+	else
+		plat_data->flags &= ~MV_PP2_PDATA_F_SGMII;
+
 	if (port < MV_UBOOT_ETH_PORTS) {
 		plat_data->mtu = mvMtu[port];
 		if (plat_data->mtu == 0) {
diff --git a/arch/arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacApi.c b/arch/arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacApi.c
index 74344b4..e32c9a4 100644
--- a/arch/arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacApi.c
+++ b/arch/arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacApi.c
@@ -102,10 +102,11 @@ void mvEthPortSgmiiSet(int port, int enable)
 	MV_U32 regVal;
 
 	regVal = MV_REG_READ(ETH_GMAC_CTRL_2_REG(port));
+
 	if (enable)
-		regVal |= ETH_GMAC_PCS_ENABLE_MASK;
+		regVal |= (ETH_GMAC_PCS_ENABLE_MASK | ETH_GMAC_INBAND_AN_MASK);
 	else
-		regVal &= ~ETH_GMAC_PCS_ENABLE_MASK;
+		regVal &= ~(ETH_GMAC_PCS_ENABLE_MASK | ETH_GMAC_INBAND_AN_MASK);
 
 	MV_REG_WRITE(ETH_GMAC_CTRL_2_REG(port), regVal);
 }
diff --git a/arch/arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacRegs.h b/arch/arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacRegs.h
index f1e014f..42177c1 100644
--- a/arch/arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacRegs.h
+++ b/arch/arm/plat-armada/mv_hal/pp2/gmac/mvEthGmacRegs.h
@@ -111,6 +111,9 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #define ETH_GMAC_CTRL_2_REG(p)             (ETH_REG_BASE(p) + 0x8)
 
+#define ETH_GMAC_INBAND_AN_BIT            0
+#define ETH_GMAC_INBAND_AN_MASK           (1 << ETH_GMAC_INBAND_AN_BIT)
+
 #define ETH_GMAC_PCS_ENABLE_BIT            3
 #define ETH_GMAC_PCS_ENABLE_MASK           (1 << ETH_GMAC_PCS_ENABLE_BIT)
 
-- 
1.7.5.4

